File: valid_parenthesis/valid_parenthesis.py
''' 
Given a string s  containing the characters '(',')', '{', '}', '[' and ']'
determine if the string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Every close bracket has a corresponding open bracket of the same type.

Example 1:

 Input S = "()"
 Output = True

Example 2:

 Input: s = "()[]{}"
 Output: True

Example 3:

 Input: s = "(]"
 Output: False
'''

def isValid(s: str) -> bool:

    # len of string has to be even
    if len(s) % 2:
        return False

    opening = {'(', '[', '{'}
    closing = {')', ']' , '}'}
    matching = { '[' : ']', '(': ')', '{': '}'}

    stack = []
    for char in s:
        # if its an opening
        if char in opening:
            stack.append(char)

        #else if its a closing bracket
        elif char in closing:
            if stack:
                opening_bracket = stack.pop()
            else:
                return False
                
            if matching[opening_bracket] != char:
                return False                    

    if stack: # if the stack is not empty then uneven 
        return False
    else:
        return True


if __name__ == "__main__":

   
    # Basic Examples
   
    #Example 1
    input = "()"
    print("Example 1: s =", input)
    print("Output:", isValid(input))
    
    #Example 2
    input = "()[]{}"
    print("Example 2: s =", input)
    print("Output:", isValid(input))

    #Example 3
    input = "(]"
    print("Example 3: s =", input)
    print("Output:", isValid(input))

File: valid_parenthesis/valid_parenthesis.cpp
/* 
Given a string s  containing the characters '(',')', '{', '}', '[' and ']'
determine if the string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Every close bracket has a corresponding open bracket of the same type.

Example 1:

 Input S = "()"
 Output = True

Example 2:

 Input: s = "()[]{}"
 Output: True

Example 3:

 Input: s = "(]"
 Output: False

*/
#include <iostream>
//#include <format>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <stack>

using namespace std;

bool isValid(string s) {

    unordered_map<char,char> mapping = { {')' , '('}, {']', '['} , {'}', '{'} };
    unordered_set<char> opening = { '(', '[', '{'};
    stack<char> stack;

    //if its uneven, fail fast
    if ( s.size() % 2 > 0 ){
        return false;
    }

    for( auto &chr : s){

        //opening bracket
        if( opening.count(chr) ){
            stack.push(chr);
        }
        //closing bracket
        else {
            if ( mapping.count(chr) && !stack.empty() ) {
                char opening_bracket = stack.top();
                stack.pop();
                //check if the opeing bracket on stack match current closing bracket
                if (mapping.at(chr) != opening_bracket){
                    return false;
                }
            }
            else {
                return false;
            }
        }
    }

    if (stack.size() > 0 ){
        return false;
    }
    else{
        return true;
    }
}

int main() {

    // Basic Examples
   
    //Example 1
    string input = "()";
    cout << "Example 1: s = " << input << '\n';
    cout << "Output: " << isValid(input) << '\n';


    //Example 2
    input = "()[]{}";
    cout << "Example 2: s = " << input << '\n';
    cout << "Output: " << isValid(input) << '\n';

    //Example 3
    input = "(]";
    cout << "Example 3: s = " << input << '\n';
    cout << "Output: " << isValid(input) << '\n';

}